nominal/network/gzip_requests_adapter.py

```diff
@@ -0,0 +1,79 @@
+from __future__ import annotations
+
+import gzip
+import io
+from typing import Any, Mapping, MutableMapping
+
+import requests
+from conjure_python_client._http.requests_client import TransportAdapter
+
+
+class GzipRequestsAdapter(TransportAdapter):
+    """Adapter used with `requests` library for sending gzip-compressed data.
+
+    Based on: https://github.com/psf/requests/issues/1753#issuecomment-417806737
+    """
+
+    ACCEPT_ENCODING = "Accept-Encoding"
+    CONTENT_ENCODING = "Content-Encoding"
+    CONTENT_LENGTH = "Content-Length"
+
+    COMPRESSION_LEVEL = 1
+
+    def add_headers(self, request: requests.PreparedRequest, **kwargs: Any) -> None:
+        """Tell the server that we support compression."""
+        super().add_headers(request, **kwargs)  # type: ignore[no-untyped-call]
+
+        body = request.body
+        if body is None:
+            return
+
+        if isinstance(body, (bytes, str)):
+            content_length = len(body)
+        else:
+            content_length = body.seek(0, 2)
+            body.seek(0, 0)
+
+        headers: MutableMapping[str, str] = {
+            self.ACCEPT_ENCODING: "gzip",
+            self.CONTENT_ENCODING: "gzip",
+            self.CONTENT_LENGTH: str(content_length),
+        }
+        request.headers.update(headers)
+
+    def send(
+        self,
+        request: requests.PreparedRequest,
+        stream: bool = False,
+        timeout: float | tuple[float, float] | tuple[float, None] | None = None,
+        verify: bool | str = True,
+        cert: bytes | str | tuple[bytes | str, bytes | str] | None = None,
+        proxies: Mapping[str, str] | None = None,
+    ) -> requests.Response:
+        """Compress data before sending."""
+        if stream:
+            # Typechecking
+            if isinstance(request.body, (bytes, str)):
+                raise ValueError("Expected request body to not be bytes or string if stream")
+            elif request.body is None:
+                raise ValueError("Expected request body to be non-null whin streaming")
+
+            # Having a file-like object, therefore we need to stream the
+            # content into a new one through the compressor.
+            compressed_body = io.BytesIO()
```

Comment:
typically a stream object _is_ a stream specifically to avoid not reading everything in memory, so probably would nix this. Most of the conjure endpoints don't use a stream anyways (only the binary types, and I think for all of those we have a multipart upload variant that the client uses) so I think we could safely just `if stream: super().send(...)`